include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -189,68 +203,124 @@ struct BFloat16_t {
+namespace detail {
+/** \brief This is a tagging template type. Use it with Base<T> to indicate that the C++ interface object
+ *   has no ownership of the underlying C object even though it is not being const.
+ */
+template <typename T>
+struct Unowned {
+  using Type = T;
+};
+
+/** \brief Used internally by the C++ API. C++ wrapper types inherit from this.
+ *   This is a zero cost abstraction to wrap the C API objects and delete them on destruction.
  *
- * This is a zero cost abstraction to wrap the C API objects and delete them on destruction.
- * There is a secondary class 'Unowned<T>' that is used to prevent deletion on destruction (Used for return types that are
- * not owned by the caller)
+ * All of the C++ classes function as
+ *  a) smart pointers that own objects that are created by the underlying C API. Their size is just
+ *  a pointer size, no need to dynamically allocate them. Use them by value.
+ *  b) convenient interfaces that return C++ objects and further enhance exception and type safety.
  *
  */
+
+/// <summary>
+/// This is a non-const pointer holder that is move-only. Disposes of the pointer on destruction.
+/// </summary>
 template <typename T>
 struct Base {
+ public:
   using contained_type = T;
 
-  Base() = default;
-  Base(T* p) : p_{p} {
-    if (!p)
-      ORT_CXX_API_THROW("Allocation failure", ORT_FAIL);
-  }
+  constexpr Base() = default;
+  constexpr explicit Base(contained_type* p) noexcept : p_{p} {}
   ~Base() { OrtRelease(p_); }
 
-  operator T*() { return p_; }
-  operator const T*() const { return p_; }
-
-  /// \brief Releases ownership of the contained pointer
-  T* release() {
+  /// \brief Relinquishes ownership of the contained C object pointer
+  /// The underlying object is not destroyed
+  contained_type* release() {
     T* p = p_;
     p_ = nullptr;
     return p;
   }
 
- protected:
   Base(const Base&) = delete;
   Base& operator=(const Base&) = delete;
+
   Base(Base&& v) noexcept : p_{v.p_} { v.p_ = nullptr; }
   void operator=(Base&& v) noexcept {
     OrtRelease(p_);
     p_ = v.release();
   }
 
-  T* p_{};
+  constexpr operator contained_type*() const noexcept { return p_; }
 
-  template <typename>
-  friend struct Unowned;  // This friend line is needed to keep the centos C++ compiler from giving an error
+ protected:
+  contained_type* p_{};
 };
 
-/** \brief Wraps an object that inherits from Ort::Base and stops it from deleting the contained pointer on destruction
- *
- * This has the effect of making it not own the memory held by Ort::Base.
- */
+/// <summary>
+/// This is a const T template specialization of Base<T>. This accepts a const pointer
+/// to the underlying C object and does not destroy it, as we do not own it.
+/// </summary>
+/// <param name="p"></param>
 template <typename T>
-struct Unowned : T {
-  Unowned(typename T::contained_type* p) : T{p} {}
-  Unowned(Unowned&& v) : T{v.p_} {}
-  ~Unowned() { this->release(); }
+struct Base<const T> {
```

Comment:
have you considered `Base<Unowned<const T>>` as an alternative to `Base<const T>`?
it seems like the "unowned" part would be more explicit that way.

also, looking at things from a smart pointer perspective, `std::unique_ptr<const T>` does have ownership.

Reply:
The specialization would be a bit more complex. We do need interfaces w/o ownership to escape the C pointers pitfall.